Initial commit.

akebrissman opened the pull request:
Fixed an initial version of the sample project for storing firebase tokens in  a database. 
I would like to have review of "The python way of doing things". 
- Code style
- Files and folder layout
- README content
- Function description yes/no? If Yes what is the best way of doing a function description. 

I also have a technical issue
The endpoint declaration in app.py is api.add_resource(Firebase, '/firebase/<string:imsi>')
This means that the endpoint must send host/firebase/12345678909876 for all verbs GET, DELETE, PUT and POST. 
In my mind POST shall not use the arguments but instead have the imsi in the json body. 

Is that correct and how can that be solved?

akebrissman committed:
Initial commit.

mblomdahl commented on .idea/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
This file should probably be removed. 🚫

mblomdahl commented on README.md:
````diff
@@ -1,2 +1,28 @@
+Install packages
+```
+pip install -r requirements.txt
+```
````
Overly verbose IMHO, copy-pasteable is better. :)

```md
### Installing

Create a Python 3.6 virtual environment and install the packages:
    
    pip install virtualenv
    virtualenv venv
    source venv/bin/activate
    pip install -r requirements.txt

```

mblomdahl commented on resources/firebase.py:
```diff
@@ -0,0 +1,74 @@
+class Firebase(Resource):
+    parser = reqparse.RequestParser()
+    parser.add_argument('imsi', type=str, required=True, help="This field canot be left blank!")
+    parser.add_argument('token', type=str, required=True, help="This field canot be left blank!")
```
`canot -> cannot`

mblomdahl commented on resources/firebase.py:
```diff
@@ -0,0 +1,74 @@
+
+        return {'message': 'Firebase item not found'}, 404
+
+    def put(self, imsi):
```
How about adding some nice type hints? E.g. `def put(self, imsi: str) -> Tule[dict, int]:`

mblomdahl commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Flask
+Flask-RESTful
+Flask-SQLAlchemy
```
Would recommend creating this using `pip freeze > requirements.txt` to avoid confusion around versions of things.

mblomdahl reviewed (changes requested):
Looking like a good start! 🥇 Do you intend to distribute it as a installable? Typically good idea IMHO. If so, good to add it from the get-go. Check out my hobby project earlier this spring for an example: https://github.com/mblomdahl/qsck/pull/2/commits/be4250a787c84fdf58990f678f91ba6cc7c46b14?file-filters%5B%5D=.py

akebrissman committed:
Changes after code review.

akebrissman committed:
Changes after code review.

akebrissman committed:
Delete .gitignore

akebrissman committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	resources/firebase.py

akebrissman committed:
Added a top folder project
Added a test folder with unit test class

akebrissman committed:
Added Setup files

carlba commented on resources/firebase.py:
```diff
@@ -0,0 +1,71 @@
+
+        fb = None
+        try:
+            fb = FirebaseModel.find_by_imsi(imsi)
```
I would enclose this in the same try except statement. If something goes wrong with the find this will silently be ignored.

mblomdahl commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+    author='Åke Brissman',
+    author_email='ake.brissman@gmail.com',
+    description='Referense project for making a RESTful application',
+
```
Add Flask, Flask-RESTful and Flask-SQLAlchemy as "install requires" dependencies here, as your did in your initial requirements.txt

akebrissman replied:
Just these three or all dependencies from requirements.txt?

mblomdahl replied:
Just the ones that you depend upon primarily (packages you would import, typically).

mblomdahl commented on app.py:
```diff
@@ -0,0 +1,29 @@
+
+@app.before_first_request
+def create_tables():
+    db.create_all()
```
Move to main block.

mblomdahl commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+    name='gateway',
+    version='0.1',
+    packages=['models', 'resources'],
+    url='https://github.com/akebrissman',
```
Should be the full URL to this repo.

mblomdahl commented on README.md:
````diff
@@ -1,2 +1,20 @@
+```
+Run
+./app.py
+```
````
About 

> * Function description yes/no? If Yes what is the best way of doing a function description.

I think the README is a good place to start. After installing and spinning it up; how does it work actually?! Then when it becomes too big and complex to present in a README, move it to a separate document. And when it's as big and complex as Passport or Analytics4, give it its own Sphinx site like http://docs.gunicorn.org 🍔

akebrissman replied:
I was more thinking of the functions/prodcedures (def ....) shall be documented. 
I guess type hinting replaces the need for that to some extent.

mblomdahl reviewed (changes requested):
Getting better every day! 🦄

akebrissman committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	setup.py

akebrissman committed:
Configured full path to the repo

akebrissman committed:
First try to add unit test for Firebase Resource and Firebase Model

akebrissman committed:
Added postman files to test the endpoints

mblomdahl commented on project/tests/test_firebase_model.py:
```diff
@@ -0,0 +1,42 @@
+        print("setUp")
+        # self.app = app.test_client()
+        db.drop_all()
+        db.create_all()
```
I've been struggling to figure out what's wrong with these test suites.. Can't figure it out. Even dug up Passport project from old archives and tried to backport the setUp/createFixtures from there. To no avail..

At this point I would recommend dropping the stdlib unittest approach and use the `pytest` library I recommended some week back. Here's an example of how it's used in conjunction with a database in my old OAuth project: https://github.com/libris/xl_auth/pull/4/commits/234f3106534102f06e57c787b03bb41b177c8ed1

It's also the framework used in `qsck`, check this initial PR adding the basic structure ("starting small"): https://github.com/mblomdahl/qsck/pull/2

Another good idea from these two repos mentioned above is to keep tests in `./tests` package, and the application source in `./gateway` directory. The "`./project`" dir with nested tests dir is a bit unfortunate.

Good things to do:

- [x] switch to pytest
- [x] make the simplest ever pytest test that passes (one not requiring database to be setup)
- [x] `./project -> ./gateway`
- [x] `./project/tests -> ./tests`

akebrissman committed:
Added Pytest to the project.

akebrissman committed:
Added Pytest to the project.

akebrissman committed:
Merge remote-tracking branch 'origin/feature/pytest' into feature/pytest

# Conflicts:
#	project/__init__.py
#	tests/conftest.py

vtepliuk committed:
Correct package structure from PR #1

- Moved dir `./project/models -> ./gateway/models`
- Moved dir `./project/resources -> ./gateway/resources`
- Moved file `./project/__init__.py -> ./gateway/__init__.py`
- Moved file `./app.py -> ./gateway/app.py`
- Moved file `./db.py -> ./gateway/db.py`
- Correct `gateway` import paths for `./tests/conftest.py`
- Corrected import paths and then commented-out not-working tests in
  `./tests`:
  - `test_firebase_model.py`
  - `test_firebase_resource.py`
  - `test_models.py`
- Added 2 simple tests in `./tests/test_the_simplest_thing.py`
- Added *Contributing* section to README, from mblomdahl/qsck#1
- Updated `setup.py` with
  - correct spelling
  - target package `gateway`
  - `setup_requires` and `install_requires` attributes

vtepliuk committed:
Fixing comments on PR #2

- Added tests that imports db and SQLAlchemy from gateway library
- Added pytest.mark.skip decorator in test_models.py instead of
  commenting it out
- Fixed direct import of SQLAlchemy

akebrissman committed:
Merge pull request #2 from mkdevops-se/feature/pytest

Correct package structure from PR #1

mblomdahl committed:
Merge pull request #3 from akebrissman/feature/pytest

feature/pytest goes home

akebrissman committed:
Restructured the project by moving the app object and the database to outside the gateway project
Updated Postman error and the initialization of flask-sql-alchemy

akebrissman committed:
Added some test cases for FirebaseModel

akebrissman committed:
Merge pull request #5 from akebrissman/test-cases-for-model

Test cases for model

akebrissman committed:
Organized the imports according to Pep-8 Import grouping.

akebrissman committed:
Improved the project by renaming flask-test.cfg to flask_cfg
Added test cases to reach 100% test coverage

akebrissman committed:
Added .travis file for tesing.

akebrissman committed:
Added shebang to the app.py file

akebrissman committed:
Merge pull request #4 from akebrissman/restructure-project

Restructure of the project
Organised the imports according to Pep-8 Import grouping
Added .travis file for testing
Added shebang to the app.py file

Commit:
Cleanup and support for Python 3.8

- changed __repr__ and __srt__ to be more in line with best practices
- Used more specific exception in the exception handler
- Updated README.md about how to create virtual env in python 3.8
- Updated packages in requirements.txt
